scratchpad/qaoa_eng_vs_be_adv.py

```diff
@@ -0,0 +1,197 @@
+    suffix = {
+        3: 'k'
+        ,6: 'M'
+        ,9: 'G'
+        , 12: 'T'
+    }[ord]
+    return f'{(flops/10**ord).round(2)}{suffix}'
+
+def obj2dict(obj):
+    keys = [x for x in dir(obj) if x[0]!='_']
+    return dict((key, obj.__getattribute__(key)) for key in keys)
+
+
+@lru_cache
+def get_gpu_props_json():
+    try:
+        import torch
+        devprops = torch.cuda.get_device_properties(torch.cuda.current_device())
+        return obj2dict(devprops)
+    except:
+        return None
+
+'''
+Main Functions:
+'''
+
+'''
+Function: Generate a single report of a backend running a problem
+I/O: backend_name, problem_descrp -> report table, byte, gen_time, num_buckets
+'''
+def gen_be_pt_report(backend_name:str , pt: list):
+    n, p, d, type = pt
+    timing = pyrofiler.timing
+    with timing(callback=lambda x: None) as gen:
+        G, gamma, beta = get_test_problem(n=n,p=p,d=d, type = type)
+        curr_backend = get_perf_backend(backend_name)
+        sim = QAOAQtreeSimulator(QtreeQAOAComposer, backend = curr_backend)
+    sim.energy_expectation(G, gamma=gamma, beta=beta)
+    curr_backend.gen_report(show = False)
+
+    tensor_dims= []
+    for x in curr_backend._profile_results:
+        bucket_signature, _ = curr_backend._profile_results[x]
+        for tensor in bucket_signature:
+            tensor_2_size = [qtreeVar.size for qtreeVar in tensor]
+            tensor_dim = np.prod(tensor_2_size)
+            tensor_dims.append(tensor_dim)
+
+    report_record = curr_backend.report_table.records
+
+    mean_record = np.mean(report_record, axis = 0)
+    max_record = np.amax(report_record, axis = 0)
+    sum_record = np.sum(report_record, axis = 0)
+    title_record = curr_backend.report_table._title_row()[1:]
+    byte = np.sum(tensor_dims)
+    gen_time = gen.result
+    num_buckets = len(curr_backend.report_table.records)
+
+    return max_record, mean_record, sum_record, title_record, byte, gen_time, num_buckets
+
+
+'''
+Function: Generate a collection of above report, and process them into final usable form
+I/O: ... -> processed data is a dict, directly usable by json
+'''
+def collect_process_be_pt_report(repeat: int, backend_name: str, pt: list):
+    means_frame = []
+    sums_frame = []
+    maxs_frame = []
+    byte_frame = []
+    gen_time_frame = []
+    bucket_num_frame = []
+    title_2_data = {}
+
+    for _ in range(repeat):
+        maxs, means, sums, titles, byte, gen_time, num_buckets = gen_be_pt_report(backend_name = backend_name, pt = pt)
+        maxs_frame.append(maxs)
+        means_frame.append(means)
+        sums_frame.append(sums)
+        byte_frame.append(byte)
+        gen_time_frame.append(gen_time)
+        bucket_num_frame.append(num_buckets)
+
+    means_frame = np.array(means_frame)
+    means_frame = np.transpose(means_frame)
+    sums_frame = np.array(sums_frame)
+    sums_frame = np.transpose(sums_frame)
+    maxs_frame = np.array(maxs_frame)
+    maxs_frame = np.transpose(maxs_frame)
```

Comment:
This part is a bit hard to understand, because it's not very clear what is the format of the data.